src/blueapi/service/authentication.py

```diff
@@ -0,0 +1,231 @@
+    @abstractmethod
+    def load_cache(cache) -> Cache | None: ...
+    @abstractmethod
+    def delete_cache(self) -> None: ...
+
+
+class SessionCacheManager(CacheManager):
+    def __init__(self, token_path: Path | None) -> None:
+        self._token_path: Path = token_path if token_path else self._get_xdg_cache_dir()
+
+    @cached_property
+    def _file_path(self) -> str:
+        return os.path.expanduser(self._token_path)
+
+    def save_cache(self, cache: Cache) -> None:
+        cache_json: str = cache.model_dump_json()
+        cache_base64: bytes = base64.b64encode(cache_json.encode("utf-8"))
+        self.delete_cache()
+        with open(self._file_path, "xb") as token_file:
+            token_file.write(cache_base64)
+        os.chmod(self._file_path, 0o600)
+
+    def load_cache(self) -> Cache | None:
+        try:
+            with open(self._file_path, "rb") as cache_file:
+                cache_base64: bytes = cache_file.read()
+                cache_json: str = base64.b64decode(cache_base64).decode("utf-8")
+                return TypeAdapter(Cache).validate_json(cache_json)
+        except Exception:
+            return None
+
+    def delete_cache(self) -> None:
+        Path(self._file_path).unlink(missing_ok=True)
+
+    def _get_xdg_cache_dir(self) -> Path:
+        """
+        Return the XDG cache directory.
+        """
+        cache_dir = os.environ.get("XDG_CACHE_HOME")
+        if not cache_dir:
+            cache_dir = os.path.expanduser(BLUEAPI_CACHE_LOCATION)
+            if cache_dir.startswith("~/"):  # Expansion failed.
+                raise ValueError("Please specify auth_token_path")
+        return Path(cache_dir)
+
+
+class SessionManager:
+    def __init__(self, server_config: OIDCConfig, cache_manager: CacheManager) -> None:
+        self._server_config = server_config
+        self._cache_manager: CacheManager = cache_manager
+
+    @classmethod
+    def from_cache(cls, auth_token_path: Path | None) -> SessionManager | None:
+        cacheManager = SessionCacheManager(auth_token_path)
+        cache = cacheManager.load_cache()
+        if cache:
+            return SessionManager(
+                server_config=cache.oidc_config, cache_manager=cacheManager
+            )
+
+    def delete_cache(self) -> None:
+        self._cache_manager.delete_cache()
+
+    def get_valid_access_token(self) -> str:
+        """
+        Retrieves a valid access token.
+
+        Returns:
+            str: A valid access token if successful.
+            "": If the operation fails (no valid token could be fetched or refreshed)
+        """
+        cache = self._cache_manager.load_cache()
+        if cache:
+            try:
+                self.decode_jwt(cache.access_token)
+                return cache.access_token
+            except jwt.ExpiredSignatureError:
+                return self._refresh_auth_token(cache.refresh_token)
+            except Exception:
+                return ""
+        return ""
```

Comment:
As above, if you remove the path where cache is None, you don't have to catch this empty string elsewhere.